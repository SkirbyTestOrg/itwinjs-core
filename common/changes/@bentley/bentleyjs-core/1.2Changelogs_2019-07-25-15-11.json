{
  "changes": [
    {
      "comment": "",
      "packageName": "@bentley/bentleyjs-core",
      "type": "none"
    }
  ],
  "packageName": "@bentley/bentleyjs-core",
  "email": "31107829+calebmshafer@users.noreply.github.com"
}