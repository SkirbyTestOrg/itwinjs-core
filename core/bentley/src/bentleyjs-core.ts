/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
export * from "./Assert";
export * from "./BeEvent";
export * from "./BentleyError";
export * from "./BeSQLite";
export * from "./Compare";
export * from "./Dictionary";
export * from "./Disposable";
export * from "./Id";
export * from "./IndexMap";
export * from "./JsonUtils";
export * from "./Logger";
export * from "./ActivityLoggingContext";
export * from "./LRUMap";
export * from "./SortedArray";
export * from "./StringUtils";
export * from "./Time";
export * from "./PriorityQueue";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("bentleyjs-core", BUILD_SEMVER);
}

/** @module Utils */

/** @docs-package-description
 * The bentleyjs-core package contains classes to solve problems that are common for both client and server use cases.
 */
/**
 * @docs-group-description BeSQLite
 * Classes for working with SQLite databases. SQLite underlies IModelDb and ECDb - see [Executing ECSQL]($docs/learning/ECSQL.md)
 */
/**
 * @docs-group-description Errors
 * Classes for working with errors.
 */
/**
 * @docs-group-description Events
 * Classes for raising and handling events.
 */
/**
 * @docs-group-description Ids
 * Classes for working with unique identifiers.
 */
/**
 * @docs-group-description Logging
 * Classes for configuring and logging diagnostic messages - see [Learning about Logging]($docs/learning/common/Logging.md)
 */
/**
 * @docs-group-description Collections
 * Specialized, customizable collection classes like priority queues.
 */
/**
 * @docs-group-description Utils
 * Miscellaneous utility classes.
 */
