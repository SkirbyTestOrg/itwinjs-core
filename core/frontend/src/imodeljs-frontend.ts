/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/

export * from "./tools/AccuDrawTool";
export * from "./tools/EditManipulator";
export * from "./tools/ElementSetTool";
export * from "./tools/EventController";
export * from "./tools/MeasureTool";
export * from "./tools/SelectTool";
export * from "./tools/ToolAdmin";
export * from "./tools/Tool";
export * from "./tools/ViewTool";
export * from "./tools/ClipViewTool";
export * from "./tools/PrimitiveTool";
export * from "./tools/IdleTool";
export * from "./tools/ToolAssistance";
export * from "./AccuDraw";
export * from "./AccuSnap";
export * from "./AuxCoordSys";
export * from "./DevTools";
export * from "./FrontendRequestContext";
export * from "./CategorySelectorState";
export * from "./ContextRealityModelState";
export * from "./SpatialClassification";
export * from "./DisplayStyleState";
export * from "./ElementLocateManager";
export * from "./EmphasizeElements";
export * from "./EntityState";
export * from "./FuzzySearch";
export * from "./GeoServices";
export * from "./HitDetail";
export * from "./IModelConnection";
export * from "./ImageUtil";
export * from "./FrontendLoggerCategory";
export * from "./Marker";
export * from "./ModelSelectorState";
export * from "./ModelState";
export * from "./NotificationManager";
export * from "./plugin/Plugin";
export * from "./plugin/PluginUiManager";
export * from "./RelativePosition";
export * from "./RenderScheduleState";
export * from "./SelectionSet";
export * from "./Sheet";
export * from "./Sprites";
export * from "./StandardView";
export * from "./SubCategoriesCache";
export * from "./TentativePoint";
export * from "./QuantityFormatter";
export * from "./ViewContext";
export * from "./ViewManager";
export * from "./Viewport";
export * from "./ViewState";
export * from "./IModelApp";
export * from "./NoRenderApp";
export * from "./tile/TileAdmin";
export * from "./tile/Tile";
export * from "./tile/TileTree";
export * from "./tile/WebMapTileTree";
export * from "./tile/BackgroundTerrainTileTree";
export * from "./tile/CesiumWorldTerrainTileTree";
export * from "./tile/BingElevation";
export * from "./render/FeatureSymbology";
export * from "./render/GraphicBuilder";
export * from "./render/MockRender";
export * from "./render/System";
export * from "./render/webgl/Target";
export * from "./RenderCompatibility";
export * from "./oidc/OidcBrowserClient";
export * from "./TerrainProvider";
export * from "./properties/Description";
export * from "./properties/EditorParams";
export * from "./properties/Record";
export * from "./properties/Value";
export * from "./properties/ToolSettingsValue";
export * from "./properties/PrimitiveTypes";
export * from "./properties/BaseQuantityDescription";
export * from "./properties/AngleDescription";
export * from "./properties/LengthDescription";
export * from "./SolarCalculate";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
declare var BUILD_TYPE: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("imodeljs-frontend", BUILD_SEMVER);
  (window as any).iModelJsVersions.set("buildType", BUILD_TYPE);
}

/** @docs-package-description
 * The ($frontend) package always runs in a web browser. It contains classes for [querying iModels and showing views]($docs/learning/frontend/index.md).
 */

/**
 * @docs-group-description IModelApp
 * Classes for configuring and administering an iModel.js application.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description IModelConnection
 * Classes for working with a connection to an [iModel briefcase]($docs/learning/IModels.md)
 */
/**
 * @docs-group-description ElementState
 * Classes for working with the *state* of Elements in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description ModelState
 * Classes for working with the *state* of Models in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Tools
 * Classes for [working with Tools]($docs/learning/frontend/Tools.md)
 */
/**
 * @docs-group-description Views
 * Classes for [working with Views]($docs/learning/frontend/Views.md)
 */
/**
 * @docs-group-description LocatingElements
 * Classes for locating and snapping to elements in views.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description AccuDraw
 * AccuDraw provides helpful assistance for creating and modifying elements in a view.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Notifications
 * Notifications provide feedback to the user of something of interest.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Plugins
 * Classes for creating and managing runtime [Plugins]($docs/learning/frontend/Plugins.md)
 */
/**
 * @docs-group-description Properties
 * Classes for working with property records and descriptions.
 */
/**
 * @docs-group-description Rendering
 * Classes for rendering the contents of views.
 */
/**
 * @docs-group-description SelectionSet
 * Classes for working with the set of selected elements.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
