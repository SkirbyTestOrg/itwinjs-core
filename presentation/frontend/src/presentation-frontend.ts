/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

export { default as PersistenceHelper } from "./PersistenceHelper";
export { default as Presentation } from "./Presentation";
export { default as PresentationManager } from "./PresentationManager";
export { default as RulesetVariablesManager } from "./RulesetVariablesManager";

/** @module UnifiedSelection */
export { default as SelectionChangeEvent, SelectionChangeEventArgs, SelectionChangeType, SelectionChangesListener } from "./selection/SelectionChangeEvent";
export { default as ISelectionProvider } from "./selection/ISelectionProvider";
export { default as SelectionManager } from "./selection/SelectionManager";
export { default as SelectionHandler } from "./selection/SelectionHandler";
