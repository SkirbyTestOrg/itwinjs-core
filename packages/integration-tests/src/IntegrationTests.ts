/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
// a hack to get imodeljs-frontend import working
(global as any).WebGLRenderingContext = require("gl"); // tslint:disable-line:no-var-requires
// common includes
import TestRpcManager from "@helpers/TestRpcManager";
// backend includes
import { IModelHost } from "@bentley/imodeljs-backend";
import { ECPresentation as ECPresentationBackend, ECPresentation } from "@bentley/ecpresentation-backend";
// frontend includes
import { StandaloneIModelRpcInterface, IModelReadRpcInterface } from "@bentley/imodeljs-common";
import { ECPresentationRpcInterface } from "@bentley/ecpresentation-common";
import { IModelApp } from "@bentley/imodeljs-frontend";
import { ECPresentation as ECPresentationFrontend } from "@bentley/ecpresentation-frontend";

process.env.NODE_ENV = "development";
let isInitialized = false;

export const initialize = () => {
  if (isInitialized)
    return;

  // install mock of browser's XMLHttpRequest for unit tests
  (global as any).XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest; // tslint:disable-line:no-var-requires

  // init backend
  IModelHost.startup();
  ECPresentationBackend.initialize({
    rulesetDirectories: ["assets/rulesets"],
    localeDirectories: ["assets/locales"],
  });

  // set up rpc interfaces
  TestRpcManager.initializeClient([StandaloneIModelRpcInterface, IModelReadRpcInterface, ECPresentationRpcInterface]);

  // init frontend
  IModelApp.startup();
  ECPresentationFrontend.initialize({
    activeLocale: IModelApp.i18n.languageList()[0],
  });

  isInitialized = true;
};

export const terminate = () => {
  if (!isInitialized)
    return;

  // terminate backend
  ECPresentation.terminate();
  IModelHost.shutdown();

  // terminate frontend
  ECPresentation.terminate();
  IModelApp.shutdown();

  isInitialized = false;
};
