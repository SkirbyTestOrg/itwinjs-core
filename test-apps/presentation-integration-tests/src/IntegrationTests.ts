/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as fs from "fs";
import * as path from "path";
import * as cpx from "cpx";
// tslint:disable-next-line:no-direct-imports
import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
// common includes
import { I18NOptions } from "@bentley/imodeljs-i18n";
import { Logger, LogLevel } from "@bentley/bentleyjs-core";
import { LoggingNamespaces } from "@bentley/presentation-common";
import { NoRenderApp } from "@bentley/imodeljs-frontend";
import { initialize as initializeTesting, terminate as terminateTesting } from "@bentley/presentation-testing";

process.env.NODE_ENV = "development";

const copyBentleyBackendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "assets");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

const copyBentleyFrontendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "public");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

class IntegrationTestsApp extends NoRenderApp {
  protected static supplyI18NOptions(): I18NOptions {
    const urlTemplate = `file://${path.resolve("lib/public/locales")}/{{lng}}/{{ns}}.json`;
    return { urlTemplate };
  }
  protected static onStartup(): void {
    cpx.copySync(`assets/**/*`, "lib/assets");
    copyBentleyBackendAssets("lib/assets");
    copyBentleyFrontendAssets("lib/public");
  }
}

export const initialize = () => {
  // init logging (enable on demand while debugging)
  Logger.initializeToConsole();
  Logger.setLevel(LoggingNamespaces.ECObjects_ECExpressions, LogLevel.None);
  Logger.setLevel(LoggingNamespaces.ECPresentation, LogLevel.None);

  initializeTesting(
    { rulesetDirectories: ["lib/assets/rulesets"], localeDirectories: ["lib/assets/locales"] },
    IntegrationTestsApp);
};

export const terminate = () => {
  terminateTesting(IntegrationTestsApp);
};
