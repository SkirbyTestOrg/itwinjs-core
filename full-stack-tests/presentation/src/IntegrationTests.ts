/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
// tslint:disable:no-direct-imports
import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import * as chai from "chai";
import * as cpx from "cpx";
import * as fs from "fs";
import * as path from "path";
import sinonChai from "sinon-chai";
import { Logger, LogLevel } from "@bentley/bentleyjs-core";
import { IModelJsConfig } from "@bentley/config-loader/lib/IModelJsConfig";
import { IModelAppOptions, NoRenderApp } from "@bentley/imodeljs-frontend";
import { I18NOptions } from "@bentley/imodeljs-i18n";
import { TestUsers } from "@bentley/oidc-signin-tool/lib/TestUsers";
import { TestUtility } from "@bentley/oidc-signin-tool/lib/TestUtility";
import { Presentation as PresentationBackend, PresentationProps as PresentationBackendProps } from "@bentley/presentation-backend";
import { LoggingNamespaces, RequestPriority } from "@bentley/presentation-common";
import { PresentationManagerProps as PresentationFrontendProps } from "@bentley/presentation-frontend";
import { initialize as initializeTesting, PresentationTestingInitProps, terminate as terminateTesting } from "@bentley/presentation-testing";

chai.use(sinonChai);

IModelJsConfig.init(true);

const copyBentleyBackendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "assets");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

const copyBentleyFrontendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "public");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

class IntegrationTestsApp extends NoRenderApp {
  protected static supplyI18NOptions(): I18NOptions {
    const urlTemplate = "file://" + path.join(path.resolve("lib/public/locales"), "{{lng}}/{{ns}}.json").replace(/\\/g, "/");
    return { urlTemplate };
  }

  public static async startup(opts?: IModelAppOptions): Promise<void> {
    await NoRenderApp.startup({ ...opts, i18n: this.supplyI18NOptions() });
    cpx.copySync(`assets/**/*`, "lib/assets");
    copyBentleyBackendAssets("lib/assets");
    copyBentleyFrontendAssets("lib/public");
  }
}

const initializeCommon = async (props: { backendTimeout?: number, useClientServices?: boolean }) => {
  // init logging
  Logger.initializeToConsole();
  Logger.setLevelDefault(LogLevel.Error);
  Logger.setLevel(LoggingNamespaces.ECObjects_ECExpressions, LogLevel.Warning);
  Logger.setLevel(LoggingNamespaces.ECPresentation, LogLevel.Warning);

  const backendInitProps: PresentationBackendProps = {
    requestTimeout: props.backendTimeout ?? 0,
    rulesetDirectories: ["lib/assets/rulesets"],
    localeDirectories: ["lib/assets/locales"],
    activeLocale: "en-PSEUDO",
    taskAllocationsMap: {
      [RequestPriority.Max]: 1,
    },
    cacheDirectory: path.join("lib", "cache"),
  };
  const frontendInitProps: PresentationFrontendProps = {
    activeLocale: "en-PSEUDO",
  };

  const frontendAppOptions: IModelAppOptions = {
    authorizationClient: props.useClientServices
      ? TestUtility.getAuthorizationClient(TestUsers.regular)
      : undefined,
  };

  const presentationTestingInitProps: PresentationTestingInitProps = {
    backendProps: backendInitProps,
    frontendProps: frontendInitProps,
    frontendApp: IntegrationTestsApp,
    frontendAppOptions,
  };

  await initializeTesting(presentationTestingInitProps);
};

export const initialize = async (backendTimeout: number = 0) => {
  await initializeCommon({ backendTimeout });
};

export const initializeWithClientServices = async () => {
  await initializeCommon({ useClientServices: true });
};

export const terminate = async () => {
  await terminateTesting();
};

export const resetBackend = () => {
  const props = PresentationBackend.initProps;
  PresentationBackend.terminate();
  PresentationBackend.initialize(props);
};
