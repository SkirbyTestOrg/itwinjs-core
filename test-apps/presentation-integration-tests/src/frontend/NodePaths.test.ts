/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import { initialize, terminate } from "../IntegrationTests";
import { OpenMode, Id64, using } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { InstanceKey, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";

describe("NodesPaths", () => {

  let imodel: IModelConnection;

  before(async () => {
    initialize();
    const testIModelName: string = "assets/datasets/Properties_60InstancesWithUrl2.ibim";
    imodel = await IModelConnection.openStandalone(testIModelName, OpenMode.Readonly);
    expect(imodel).is.not.null;
  });

  after(async () => {
    await imodel.closeStandalone();
    terminate();
  });

  it("gets filtered node paths", async () => {
    const ruleset: Ruleset = require("../../test-rulesets/NodePaths/getFilteredNodePaths");
    /* Hierarchy in the ruleset:
    filter r1
      filter ch1
      other ch2
      other ch3
        filter ch4
    other r2
    other r3
      other ch5
      filter ch6
    */
    await using(await Presentation.presentation.rulesets().add(ruleset), async () => {
      const result = await Presentation.presentation.getFilteredNodePaths({ imodel, rulesetId: ruleset.id }, "filter");
      expect(result).to.matchSnapshot();
    });
  });

  it("gets node paths", async () => {
    const ruleset: Ruleset = require("../../test-rulesets/NodePaths/getNodePaths");
    /*
    [BisCore:RepositoryModel] 0x1
      [BisCore:Subject] 0x1
        [BisCore:DefinitionPartition] ECClassGroupingNode
          [BisCore:DefinitionPartition] 0x10
        [BisCore:LinkPartition] ECClassGroupingNode
          [BisCore:LinkPartition] 0xe
    */
    await using(await Presentation.presentation.rulesets().add(ruleset), async () => {
      const key1: InstanceKey = { id: Id64.fromString("0x1"), className: "BisCore:RepositoryModel" };
      const key2: InstanceKey = { id: Id64.fromString("0x1"), className: "BisCore:Subject" };
      const key3: InstanceKey = { id: Id64.fromString("0x10"), className: "BisCore:DefinitionPartition" };
      const key4: InstanceKey = { id: Id64.fromString("0xe"), className: "BisCore:LinkPartition" };
      const keys: InstanceKey[][] = [[key1, key2, key3], [key1, key2, key4]];

      const result = await Presentation.presentation.getNodePaths({ imodel, rulesetId: ruleset.id }, keys, 1);
      expect(result).to.matchSnapshot();
    });
  });

});
