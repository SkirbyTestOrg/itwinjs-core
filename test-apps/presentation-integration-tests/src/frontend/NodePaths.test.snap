// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodesPaths gets filtered node paths 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "filteringData": Object {
          "childrenOccurances": 0,
          "occurances": 1,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "d849e52eb605991e839a5e6aa299a4b9",
              "aa959b85dd4402b50d79c57d8a6f5215",
            ],
            "type": "nodeType",
          },
          "label": "filter ch1",
        },
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "filteringData": Object {
              "childrenOccurances": 0,
              "occurances": 1,
            },
            "index": 0,
            "isMarked": false,
            "node": Object {
              "description": "description",
              "imageId": "imageId",
              "key": Object {
                "pathFromRoot": Array [
                  "d849e52eb605991e839a5e6aa299a4b9",
                  "79f473dd6d221d37d0928acdc2cb5b3c",
                  "d7abfd31da4a1dc5cc4d0307f8aa1a8e",
                ],
                "type": "nodeType",
              },
              "label": "filter ch4",
            },
          },
        ],
        "filteringData": Object {
          "childrenOccurances": 1,
          "occurances": 0,
        },
        "index": 1,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "hasChildren": true,
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "d849e52eb605991e839a5e6aa299a4b9",
              "79f473dd6d221d37d0928acdc2cb5b3c",
            ],
            "type": "nodeType",
          },
          "label": "other ch3",
        },
      },
    ],
    "filteringData": Object {
      "childrenOccurances": 2,
      "occurances": 1,
    },
    "index": 0,
    "isMarked": false,
    "node": Object {
      "description": "description",
      "hasChildren": true,
      "imageId": "imageId",
      "key": Object {
        "pathFromRoot": Array [
          "d849e52eb605991e839a5e6aa299a4b9",
        ],
        "type": "nodeType",
      },
      "label": "filter r1",
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "filteringData": Object {
          "childrenOccurances": 0,
          "occurances": 1,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "description": "description",
          "imageId": "imageId",
          "key": Object {
            "pathFromRoot": Array [
              "6da3c696f564a4aa1c913cdfb7774b51",
              "239b0de6eee1bbd89d6eb4c8affcc76f",
            ],
            "type": "nodeType",
          },
          "label": "filter ch6",
        },
      },
    ],
    "filteringData": Object {
      "childrenOccurances": 1,
      "occurances": 0,
    },
    "index": 1,
    "isMarked": false,
    "node": Object {
      "description": "description",
      "hasChildren": true,
      "imageId": "imageId",
      "key": Object {
        "pathFromRoot": Array [
          "6da3c696f564a4aa1c913cdfb7774b51",
        ],
        "type": "nodeType",
      },
      "label": "other r3",
    },
  },
]
`;

exports[`NodesPaths gets node paths 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "filteringData": Object {
                  "childrenOccurances": 0,
                  "occurances": 0,
                },
                "index": 0,
                "isMarked": false,
                "node": Object {
                  "imageId": "ECInstanceImage://BisCore:DefinitionPartition",
                  "key": Object {
                    "instanceKey": Object {
                      "className": "BisCore:DefinitionPartition",
                      "id": "0x10",
                    },
                    "pathFromRoot": Array [
                      "4e074cc7a6fd65b2a738d22f8b912299",
                      "7f68095dd510724c73d10956820f7921",
                      "bd4972ffb7cd9bc2e1493436df5c921b",
                      "5ae7367cae801dda838b167a98c08320",
                      "f58b45a0b1dcc1005d366672a876535f",
                    ],
                    "type": "ECInstanceNode",
                  },
                  "label": "Definition Partition-0-G",
                },
              },
            ],
            "filteringData": Object {
              "childrenOccurances": 0,
              "occurances": 0,
            },
            "index": 0,
            "isMarked": false,
            "node": Object {
              "description": "A Definition Partition element indicates that there is a definition-related modeling perspective within the overall DgnDb information hierarchy. A Definition Partition is always parented to a Subject and broken down by a Definition Model.",
              "hasChildren": true,
              "imageId": "ECClassImage://BisCore:DefinitionPartition",
              "key": Object {
                "className": "BisCore:DefinitionPartition",
                "groupedInstancesCount": 1,
                "pathFromRoot": Array [
                  "4e074cc7a6fd65b2a738d22f8b912299",
                  "7f68095dd510724c73d10956820f7921",
                  "bd4972ffb7cd9bc2e1493436df5c921b",
                  "5ae7367cae801dda838b167a98c08320",
                ],
                "type": "ECClassGroupingNode",
              },
              "label": "Definition Partition",
            },
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "filteringData": Object {
                  "childrenOccurances": 0,
                  "occurances": 0,
                },
                "index": 0,
                "isMarked": true,
                "node": Object {
                  "imageId": "ECInstanceImage://BisCore:LinkPartition",
                  "key": Object {
                    "instanceKey": Object {
                      "className": "BisCore:LinkPartition",
                      "id": "0xe",
                    },
                    "pathFromRoot": Array [
                      "4e074cc7a6fd65b2a738d22f8b912299",
                      "7f68095dd510724c73d10956820f7921",
                      "bd4972ffb7cd9bc2e1493436df5c921b",
                      "ab9d6344c1e60c3a68b8dea279224873",
                      "2befd7b22098b671c658a3d15b904677",
                    ],
                    "type": "ECInstanceNode",
                  },
                  "label": "Link Partition-0-E",
                },
              },
            ],
            "filteringData": Object {
              "childrenOccurances": 0,
              "occurances": 0,
            },
            "index": 1,
            "isMarked": false,
            "node": Object {
              "description": "A Link Partition element indicates that there is a link-related modeling perspective within the overall DgnDb information hierarchy. A Link Partition is always parented to a Subject and broken down by a Link Model.",
              "hasChildren": true,
              "imageId": "ECClassImage://BisCore:LinkPartition",
              "key": Object {
                "className": "BisCore:LinkPartition",
                "groupedInstancesCount": 1,
                "pathFromRoot": Array [
                  "4e074cc7a6fd65b2a738d22f8b912299",
                  "7f68095dd510724c73d10956820f7921",
                  "bd4972ffb7cd9bc2e1493436df5c921b",
                  "ab9d6344c1e60c3a68b8dea279224873",
                ],
                "type": "ECClassGroupingNode",
              },
              "label": "Link Partition",
            },
          },
        ],
        "filteringData": Object {
          "childrenOccurances": 0,
          "occurances": 0,
        },
        "index": 0,
        "isMarked": false,
        "node": Object {
          "hasChildren": true,
          "imageId": "ECInstanceImage://BisCore:Subject",
          "key": Object {
            "instanceKey": Object {
              "className": "BisCore:Subject",
              "id": "0x1",
            },
            "pathFromRoot": Array [
              "4e074cc7a6fd65b2a738d22f8b912299",
              "7f68095dd510724c73d10956820f7921",
              "bd4972ffb7cd9bc2e1493436df5c921b",
            ],
            "type": "ECInstanceNode",
          },
          "label": "Subject-0-1",
        },
      },
    ],
    "filteringData": Object {
      "childrenOccurances": 0,
      "occurances": 0,
    },
    "index": 0,
    "isMarked": false,
    "node": Object {
      "hasChildren": true,
      "imageId": "ECInstanceImage://BisCore:RepositoryModel",
      "key": Object {
        "instanceKey": Object {
          "className": "BisCore:RepositoryModel",
          "id": "0x1",
        },
        "pathFromRoot": Array [
          "4e074cc7a6fd65b2a738d22f8b912299",
          "7f68095dd510724c73d10956820f7921",
        ],
        "type": "ECInstanceNode",
      },
      "label": "Repository Model-0-1",
    },
  },
]
`;
